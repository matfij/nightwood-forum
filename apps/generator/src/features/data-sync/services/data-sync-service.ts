import * as path from 'path';
import { Worker, isMainThread } from 'worker_threads';
import fs from 'fs-extra';
import { SyncProjectDataParams } from '../models/sync-project-data-params';
import { ContentBlock } from '../../generator/models/content-block';
import { ProjectRepository } from '../../projects/data-access/project-repository';
import { PROJECT_CONTENT_PATH } from '../../generator/utils/cache-paths';

export class DataSyncService {
    constructor(private projectRepository: ProjectRepository) {}

    async syncProjectData(params: SyncProjectDataParams): Promise<void> {
        const project = await this.projectRepository.findOne({ id: params.projectId });
        if (!project || !project.id) {
            throw new Error('Project not found');
        }
        if (project.userId !== params.userId) {
            throw new Error('Access denied');
        }
        if (isMainThread) {
            const workerPath = path.resolve(__dirname, './sync-data-worker.ts');
            const worker = new Worker(workerPath);
            worker.postMessage(project);
            return new Promise((resolve, reject) => {
                worker.on('message', resolve);
                worker.on('error', reject);
            });
        }
    }

    async getSyncedProjectData(projectId: string): Promise<ContentBlock[]> {
        if (!fs.existsSync(PROJECT_CONTENT_PATH(projectId))) {
            throw new Error('Project data not synced');
        }
        const data = fs.readFileSync(PROJECT_CONTENT_PATH(projectId), { encoding: 'utf-8' });
        const blocks = JSON.parse(data);
        return blocks;
    }
}
