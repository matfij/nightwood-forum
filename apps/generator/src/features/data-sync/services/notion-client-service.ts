import { Client, isFullBlock } from '@notionhq/client';
import { NOTION_API_URL, NOTION_API_VERSION } from '../../../common/config';
import { BlockObjectResponse } from '@notionhq/client/build/src/api-endpoints';
import { ApiError, ApiErrorCode, ApiErrorName } from '../../../common/errors/api-error';

export class NotionClientService {
    private static readonly notionClient = new Client({
        baseUrl: NOTION_API_URL,
        notionVersion: NOTION_API_VERSION,
    });

    static async readPageBlocks(pageId: string, authToken: string): Promise<BlockObjectResponse[]> {
        try {
            const pageBlocks = await this.readNestedBlocks(pageId, authToken);
            return pageBlocks;
        } catch (error) {
            throw new ApiError(
                ApiErrorName.NotionConnectionFailed,
                ApiErrorCode.BadRequest,
                `Failed to retrieve data from Notion: ${error}`,
                true,
            );
        }
    }

    private static async readNestedBlocks(blockId: string, authToken: string): Promise<BlockObjectResponse[]> {
        const resultBlocks: BlockObjectResponse[] = [];
        const blocks = await this.notionClient.blocks.children.list({
            block_id: blockId,
            auth: authToken,
        });
        for (let block of blocks.results) {
            if (!isFullBlock(block)) {
                continue;
            }
            if (block.has_children) {
                const nestedBlocks = await this.readNestedBlocks(block.id, authToken);
                resultBlocks.push(block, ...nestedBlocks);
                continue;
            }
            resultBlocks.push(block);
        }
        return resultBlocks;
    }
}
