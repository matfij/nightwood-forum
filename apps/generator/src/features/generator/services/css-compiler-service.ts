import { ProjectConfig } from '../../projects/models/project-config';

export class CssCompilerService {
    static compile(chunk: string, config: ProjectConfig): string {
        for (const [key, val] of Object.entries(config)) {
            if (typeof val === 'object') {
                for (const [_key, _val] of Object.entries(val)) {
                    const fullKey = `#${key}.${_key}`;
                    chunk = chunk.replace(fullKey, `${_val}`);
                }
                continue;
            }
            const fullKey = `#${key}`;
            chunk = chunk.replace(fullKey, `${val}`);
        }
        return chunk;
    }
}
