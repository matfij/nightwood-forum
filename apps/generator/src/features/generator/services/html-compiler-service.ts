import { ContentBlock, ContentBlockType } from '../models/content-block';

export class HtmlCompilerService {
    static compile(blocks: ContentBlock[]): string {
        let content = '';
        blocks.forEach((block) => {
            content += `\n${this.compileBlock(block)}`;
        });
        return content;
    }

    private static compileBlock(block: ContentBlock): string {
        let content = '';
        let childrenContent = block.children.reduce((prev, curr) => prev + `\n${this.compileBlock(curr)}`, '');
        switch (block.type) {
            case ContentBlockType.Heading: {
                content += `<h1 class="heading-block">${block.textContent}</h1>`;
                break;
            }
            case ContentBlockType.Paragraph: {
                content += `<p class="paragraph-block">${block.textContent}${childrenContent}</p>`;
                break;
            }
            case ContentBlockType.Divider: {
                content += `<hr class="divider-block" />`;
                break;
            }
            case ContentBlockType.Callout: {
                content += `<div class="callout-block">üéâ ${block.textContent}${childrenContent}</div>`;
                break;
            }
            case ContentBlockType.ListItem: {
                content += `<li class="list-item-block">${block.textContent}<ul>${childrenContent}</ul></li>`;
                break;
            }
            case ContentBlockType.TodoItem: {
                content += `<p class="todo-block">${block.checked ? '‚úì' : '‚òê'} ${block.textContent}${childrenContent}</p>`;
                break;
            }
            case ContentBlockType.Image: {
                content += `<img class="image-block" src="${block.filePath}">`;
                break;
            }
            case ContentBlockType.File: {
                content += `<a class="file-block" href="${block.filePath}" download>üìÅ</a>`;
                break;
            }
            default: {
                content += `<p>**Unsupported block (${block.textContent})**</p>`;
                break;
            }
        }
        return content;
    }
}
