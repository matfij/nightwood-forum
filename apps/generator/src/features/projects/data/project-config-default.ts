import { ProjectConfig } from '../models/project-config';

export const PROJECT_CONFIG_DEFAULT: ProjectConfig = {
    backgroundColor: '#eed',
    fontColor: '#111',
    fontUrl: `@import url('https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500;700&display=swap');`,
    fontFamily: `'Roboto', sans-serif;`,
    heading: {
        fontSize: '24px',
        fontWeight: 'bold',
        margin: '12px auto',
    },
    paragraph: {
        fontSize: '14px',
        fontWeight: 'regular',
        margin: '0px auto',
    },
};
