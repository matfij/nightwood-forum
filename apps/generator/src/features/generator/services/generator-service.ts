import fs from 'fs';
import { Transform } from 'stream';
import { GenerateParams } from '../models/generate-params';
import { ProjectModel } from '../../projects/models/project-model';
import { WebsiteCompilerService } from './website-compiler-service';
import { DataSyncService } from './data-sync-service';

export class GeneratorService {
    static async generateWebsite(params: GenerateParams): Promise<Transform> {
        const project = await ProjectModel.findOne({ id: params.projectId });
        if (!project) {
            throw new Error('Project not found');
        }
        if (project.userId !== params.userId) {
            throw new Error('Access denied');
        }
        const projectBlocks = await DataSyncService.getSyncedProjectData(params.projectId);
        const content = WebsiteCompilerService.compile(projectBlocks);
        const htmlTemplateStream = fs.createReadStream('src/features/generator/services/templates/index.html');
        return htmlTemplateStream.pipe(
            new Transform({
                transform(chunk, encoding, callback) {
                    const replaceChunk = chunk
                        .toString()
                        .replace('#project-name#', project.notionName)
                        .replace('#project-content#', content);
                    this.push(replaceChunk);
                    callback();
                },
            }),
        );
    }
}
