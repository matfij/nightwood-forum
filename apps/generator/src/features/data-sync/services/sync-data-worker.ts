import fs from 'fs-extra';
import { parentPort } from 'worker_threads';
import { PROJECT_ASSETS_PATH, PROJECT_CACHE_PATH, PROJECT_CONTENT_PATH } from '../../generator/utils/cache-paths';
import { NotionClientService } from './notion-client-service';
import { NotionParserService } from './notion-parser-service';
import { Project } from '../../projects/models/project-model';

parentPort?.on('message', async (projectData: { _doc: Project }) => {
    try {
        const { _doc: project } = projectData;
        if (!project || !project.id || !project.notionId || !project.notionAccessCode) {
            parentPort?.postMessage({ success: false });
            return;
        }
        fs.removeSync(PROJECT_CACHE_PATH(project.id));
        if (!fs.existsSync(PROJECT_CACHE_PATH(project.id))) {
            fs.mkdirSync(PROJECT_CACHE_PATH(project.id));
        }
        if (!fs.existsSync(PROJECT_ASSETS_PATH(project.id))) {
            fs.mkdirSync(PROJECT_ASSETS_PATH(project.id));
        }
        const notionBlocks = await NotionClientService.readPageBlocks(project.notionId, project.notionAccessCode);
        const parsedBlocks = await NotionParserService.parseBlocks(project.id, notionBlocks);
        fs.outputFileSync(PROJECT_CONTENT_PATH(project.id), JSON.stringify(parsedBlocks), {
            encoding: 'utf-8',
        });
        parentPort?.postMessage({ success: true });
    } catch (err) {
        parentPort?.postMessage({ success: false });
    }
});
