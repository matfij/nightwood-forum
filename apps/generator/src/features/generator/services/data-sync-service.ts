import fs from 'fs';
import { SyncProjectDataParams } from '../models/sync-project-data-params';
import { NotionClientService } from './notion-client-service';
import { NotionParserService } from './notion-parser-service';
import { ContentBlock } from '../models/content-block';
import { ProjectRepository } from '../../projects/data-access/project-repository';

export class DataSyncService {
    PROJECT_CACHE_PATH = (projectId: string) => `.cache/proj_${projectId}.json`;

    constructor(private projectRepository: ProjectRepository) {}

    async syncProjectData(params: SyncProjectDataParams): Promise<void> {
        const project = await this.projectRepository.findOne({ id: params.projectId });
        if (!project) {
            throw new Error('Project not found');
        }
        if (project.userId !== params.userId) {
            throw new Error('Access denied');
        }
        const notionBlocks = await NotionClientService.readPageBlocks(project.notionId, project.notionAccessCode);
        const parsedBlocks = await NotionParserService.parseBlocks(notionBlocks);
        fs.writeFileSync(this.PROJECT_CACHE_PATH(params.projectId), JSON.stringify(parsedBlocks), {
            encoding: 'utf-8',
        });
    }

    async getSyncedProjectData(projectId: string): Promise<ContentBlock[]> {
        if (!fs.existsSync(this.PROJECT_CACHE_PATH(projectId))) {
            throw new Error('Project data not synced');
        }
        const data = fs.readFileSync(this.PROJECT_CACHE_PATH(projectId), { encoding: 'utf-8' });
        const blocks = JSON.parse(data);
        return blocks;
    }
}
