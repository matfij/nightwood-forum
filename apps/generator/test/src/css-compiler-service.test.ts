import { CssCompilerService } from '../../src/features/generator/services/css-compiler-service';
import { PROJECT_CONFIG_DEFAULT } from '../../src/features/projects/data/project-config-default';
import { CSS_CHUNK, CSS_CHUNK_UNSUPPORTED } from '../stubs/css-chunk';

describe('CssCompilerService', () => {
    it('should compile a valid project config', () => {
        const chunk = CSS_CHUNK;
        const config = PROJECT_CONFIG_DEFAULT;

        const parsedChunk = CssCompilerService.compile(chunk, config);

        expect(parsedChunk).toMatchSnapshot();
    });

    it('should not replace unsupported tags', () => {
        const chunk = CSS_CHUNK_UNSUPPORTED;
        const config = PROJECT_CONFIG_DEFAULT;

        const parsedChunk = CssCompilerService.compile(chunk, config);

        expect(parsedChunk).toMatchSnapshot();
    });
});
