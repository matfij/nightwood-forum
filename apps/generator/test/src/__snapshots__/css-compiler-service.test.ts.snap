// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CssCompilerService should compile a valid project config 1`] = `
"
@import url('https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500;700&display=swap');
body {
    background: #eed;
    font-family: 'Roboto', sans-serif;;
    color: #111;
}
.heading-block {
    width: 100%;
    margin: 12px auto;
    font-size: 24px;
    font-weight: bold;
}
.paragraph-block {
    width: 100%;
    margin: 0px auto;
    font-size: 14px;
    font-weight: regular;
}
"
`;

exports[`CssCompilerService should not replace unsupported tags 1`] = `
"
@import url('https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500;700&display=swap');
body {
    background: #eed;
    font-family: 'Roboto', sans-serif;;
    color: #111;
}
.unsupported-block {
    width: 100%;
    margin: #unsupported.margin;
    font-size: #unsupported.fontSize;
    font-weight: #unsupported.fontWeight;
}
"
`;
